### PR TITLE
Add libdevice bindings for numpy math functions

### LINKED ISSUE #59: test_numpy_function[arcsin] crashes omniscidb server when CUDA enabled
```
2020-04-14T11:29:49.926462 F 29925 NvidiaKernel.cpp:83 Check failed: cuLinkComplete(link_state, &cubin, &cubinSize) == CUDA_SUCCESS (999 == 0) 
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Fixes #59, #60, #71 

Depends on https://github.com/xnd-project/rbc/pull/198 [rbc] and https://github.com/omnisci/omniscidb-internal/pull/4870 [omniscidb]

### CODE REVIEW

**Comment by guilhermeleobas** [comment]
Tests are failing because `numba.cuda.mathimpl.py` was introduced two months ago, before the latest numba release. We may need to wait numba team to release Numba 0.51.0 before merging this PR.

**Comment by guilhermeleobas** [comment]
Also, the cuda target now supports almost all numpy math functions 
```
========================================================================================================== test session starts ==========================================================================================================
platform linux -- Python 3.8.6, pytest-6.1.1, py-1.9.0, pluggy-0.13.1 -- /home/guilhermel/miniconda3/envs/rbc/bin/python
cachedir: .pytest_cache
rootdir: /home/guilhermel/git/rbc
collecting ...  OmnisciDB version (5, 5, 0, 'dev-20201017-79b80643b2')
collected 89 items

rbc/tests/test_omnisci_math.py::test_numpy_function[absolute] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[conjugate] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[conj] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[fabs] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[fmax] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[fmin] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[maximum] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[minimum] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[positive] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[negative] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[sign] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[reciprocal] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[add] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[subtract] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[multiply] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[divide] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[true_divide] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[floor_divide] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[power] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[float_power] SKIPPED
rbc/tests/test_omnisci_math.py::test_numpy_function[square] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[sqrt] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[cbrt] SKIPPED
rbc/tests/test_omnisci_math.py::test_numpy_function[remainder] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[fmod] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[modf] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[modi] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[divmod0] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[sin] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[cos] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[tan] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arcsin] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arccos] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arctan] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arctan2] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[hypot] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[radians] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[rad2deg] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[deg2rad] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[degrees] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[sinh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[cosh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[tanh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arcsinh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arccosh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[arctanh] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[exp] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[expm1] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[exp2] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[log] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[log10] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[log2] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[log1p] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logaddexp] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logaddexp2] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[ldexp] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[frexp0] SKIPPED
rbc/tests/test_omnisci_math.py::test_numpy_function[around] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[round2] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[floor] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[ceil] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[trunc] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[rint] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[spacing] SKIPPED
rbc/tests/test_omnisci_math.py::test_numpy_function[nextafter] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[gcd] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[lcm] SKIPPED
rbc/tests/test_omnisci_math.py::test_numpy_function[right_shift] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[left_shift] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[heaviside] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[copysign] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[invert] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[bitwise_or] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[bitwise_xor] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[bitwise_and] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[isfinite] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[isinf] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[isnan] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[signbit] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[less] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[less_equal] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[greater] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[greater_equal] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[equal] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[not_equal] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logical_or] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logical_xor] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logical_and] PASSED
rbc/tests/test_omnisci_math.py::test_numpy_function[logical_not] PASSED

======================================================================================================== short test summary info ========================================================================================================
SKIPPED [1] rbc/tests/test_omnisci_math.py:322: Numba does not support float_power
SKIPPED [1] rbc/tests/test_omnisci_math.py:322: Numba does not support cbrt
SKIPPED [1] rbc/tests/test_omnisci_math.py:356: Use of unsupported NumPy function 'numpy.frexp' or unsupported use of the function.
SKIPPED [1] rbc/tests/test_omnisci_math.py:359: spacing: FIXME
SKIPPED [1] rbc/tests/test_omnisci_math.py:363: lcm: crashes CUDA enabled omniscidb server [rbc issue 71]
============================================================================================== 84 passed, 5 skipped, 61 warnings in 58.19s ==============================================================================================
```

File: rbc/tests/test_omnisci_math.py [hunk 1]
```diff
@@ -355,27 +356,22 @@ def test_numpy_function(omnisci, fn_name, signature, np_func):
...
-        pytest.skip(f'{fn_name}: crashes CUDA enabled omniscidb server'
-                    ' [rbc issue 59]')
-
-    if omnisci.has_cuda and fn_name in ['logaddexp']:
-        # https://github.com/xnd-project/rbc/issues/60
-        pytest.skip(f'{fn_name}: crashes CUDA enabled omniscidb server'
-                    ' [rbc issue 60]')
-
-    if omnisci.has_cuda and fn_name in ['lcm', 'ldexp']:
+    if omnisci.has_cuda and fn_name in ['lcm']:
```
Comment (guilhermeleobas) [inline rbc/tests/test_omnisci_math.py:362]: This check can be removed after #201 is merged

Comment (pearu) [reply rbc/tests/test_omnisci_math.py:362]: #201 is merged.

**Comment by pearu** [comment]
@guilhermeleobas , merging #201 required fixes that exist in this PR as well (that I notice after the merge), hence the conflicts. Could you resolve these?

File: rbc/irtools.py [hunk 2]
```diff
@@ -162,7 +161,50 @@ def init(self):
...
+            self.install_registry(npydecl.registry)
+            self.install_registry(cudamath.registry)
+            self.install_registry(libdevicedecl.registry)
+        super().load_additional_registries()
+
+# ---------------------------------------------------------------------------
+# Code generation methods
+
+
+class RemoteGPUTargetContext(cpu.CPUContext):
```
Comment (pearu) [inline rbc/irtools.py:192]: The classes RemoteGPUTypingContext and RemoteGPUTargetContext are defined twice in this file.
Cleanup?

File: rbc/omnisci_backend/__init__.py [hunk 3]
```diff
@@ -1,4 +1,6 @@
 from .numpy_funcs import *  # noqa: F401, F403
+from .npy_mathimpl import *  # noqa: F401, F403
+# from .cuda_npyimpl import *  # noqa: F401, F403
```
Comment (pearu) [inline rbc/omnisci_backend/__init__.py:3]: Remove comment.

File: rbc/omnisci_backend/cuda_npyimpl.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+registry = Registry()
+lower = registry.lower
+
+
+booleans = []
+booleans += [('isnand', 'isnanf', np.isnan)]
+booleans += [('isinfd', 'isinff', np.isinf)]
+booleans += [('isfinited', 'finitef', np.isfinite)]
+booleans += [('isnand', 'isnanf', np.isnat)]
+
```
Comment (pearu) [inline rbc/omnisci_backend/cuda_npyimpl.py:16]: Insert comment here about why some lines below are commented out.

**Review by pearu — CHANGES_REQUESTED** [review]
LGTM, there are a couple of places that require code cleanup.

General note: currently, we check numba version in tests in order to skip certain tests that are known to fail (crash the server) with older numba versions. This approach is problematic because users will be still able to crash the omniscidb server with their code. So I think we should start moving such checks to rbc code generation part so that the failures discussed above would be catched earlier and raised in the rbc side rather than at the server side (which are often fatal).

File: rbc/omnisci_backend/cuda_npyimpl.py (hunk 4 above)
Comment (guilhermeleobas) [reply rbc/omnisci_backend/cuda_npyimpl.py:21]: Done!

**Comment by guilhermeleobas** [comment]
Hi Pearu, I've addressed all your comments.

**Review by pearu — APPROVED** [review]
In the future, the behavior implemented here may need to be changed: instead of raising ForbiddenIntrinsicError, one should skip generating gpu code if it is known to fail in the server because the server would be successful in executing the corresponding cpu code.
See https://github.com/xnd-project/rbc/issues/208.

But let's get this merged and improve handling of unsupported features in future.

**Pull request merged**